analytics/pubstack/pubstack_module.go

```diff
@@ -0,0 +1,246 @@
+package pubstack
+
+import (
+	"fmt"
+	"github.com/prebid/prebid-server/analytics/pubstack/eventchannel"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/analytics/pubstack/helpers"
+
+	"github.com/prebid/prebid-server/analytics"
+)
+
+type Configuration struct {
+	ScopeId  string          `json:"scopeId"`
+	Endpoint string          `json:"endpoint"`
+	Features map[string]bool `json:"features"`
+}
+
+// routes for events
+const (
+	auction    = "auction"
+	cookieSync = "cookiesync"
+	amp        = "amp"
+	setUID     = "setuid"
+	video      = "video"
+)
+
+type bufferConfig struct {
+	timeout time.Duration
+	count   int64
+	size    int64
+}
+
+type PubstackModule struct {
+	eventChannels map[string]*eventchannel.EventChannel
+	configCh      chan *Configuration
+	scope         string
+	cfg           *Configuration
+	buffsCfg      *bufferConfig
+	mux           sync.Mutex
+}
+
+func NewPubstackModule(scope, endpoint, configRefreshDelay string, maxEventCount int, maxByteSize, maxTime string) (analytics.PBSAnalyticsModule, error) {
+	glog.Infof("[pubstack] Initializing module scope=%s endpoint=%s\n", scope, endpoint)
+
+	// parse args
+
+	refreshDelay, err := time.ParseDuration(configRefreshDelay)
+	if err != nil {
+		return nil, fmt.Errorf("fail to parse the module args, arg=analytics.pubstack.configuration_refresh_delay, :%v", err)
+	}
+
+	bufferCfg, err := newBufferConfig(maxEventCount, maxByteSize, maxTime)
+	if err != nil {
+		return nil, fmt.Errorf("fail to parse the module args, arg=analytics.pubstack.buffers, :%v", err)
+	}
+
+	defaultFeatures := make(map[string]bool)
+	defaultFeatures[auction] = false
+	defaultFeatures[video] = false
+	defaultFeatures[amp] = false
+	defaultFeatures[cookieSync] = false
+	defaultFeatures[setUID] = false
+
+	defaultConfig := &Configuration{
+		ScopeId:  scope,
+		Endpoint: endpoint,
+		Features: defaultFeatures,
+	}
+
+	pb := PubstackModule{
+		scope:         scope,
+		cfg:           defaultConfig,
+		buffsCfg:      bufferCfg,
+		configCh:      make(chan *Configuration),
+		eventChannels: make(map[string]*eventchannel.EventChannel),
+	}
+	endCh := make(chan os.Signal)
+	signal.Notify(endCh, os.Interrupt, syscall.SIGTERM)
+
+	go pb.setup()
+	go pb.start(refreshDelay, endCh)
+
+	glog.Info("[pubstack] Pubstack analytics configured and ready")
+	return &pb, nil
+}
+
+func (p *PubstackModule) LogAuctionObject(ao *analytics.AuctionObject) {
+	p.mux.Lock()
```

Comment:
Is this lock protecting the config via `isFeatureEnable` or the Push via `Push(payload)`.. or both? I'm asking because if it's just one or the either, we can restrict it's scope to minimize the lock time.

Reply:
Both, because if the configuration is updated, the flag and the map(push func) could have a new value.
Now, I'm using the RWMutex to improve the lock